### ISSUE TITLE
Retrieve row location in input file

### ISSUE THREAD

**Issue by Lawouach** [open-policy-agent/opa#2935]
##  Expected Behavior

I wish I could know which line triggered a policy in an input file.

## Actual Behavior

I don't seem to find that information.

As I understand it, https://github.com/open-policy-agent/opa/blob/db030e0b143915d5c79ebfddd80054b5b726194e/ast/location/location.go#L11 is the location in the Rego file, not the file under validation.

It's fairly common to have large Kubernetes YAML files with many resources described. Not being able to quickly know which line in that file is guilty is a pain.

Is there a way to retrieve that information programmatically?

**Comment by srenatus** [comment]
> Is there a way to retrieve that information programmatically?

You're correct about `ast.Location` being about the rego, not input or data. 🤔 

Due to the way that `input` is consumed as JSON or YAML, without any source location data attached to it, I think there's currently no way to do what you're trying to achieve.

👉 As a *workaround*, however, you could extend your result message, or object, to include some information about the resource that has triggered it. Happy to give you a pointer if you want to share some of your policy code 😃

**Comment by Lawouach** [comment]
> Due to the way that input is consumed as JSON or YAML, without any source location data attached to it, I think there's currently no way to do what you're trying to achieve.

Bummer but understable.

I haven't looked deeply into the code but I wonder if I could swap your json reader for one that keeps line numbers as it reads. Just a thought. Or, if we could get the jsonpath for a particular source line that triggered a rule.

> As a workaround, however, you could extend your result message, or object, to include some information about the resource that has triggered it. Happy to give you a pointer if you want to share some of your policy code.

Appreciated! My policies aren't particularly complicated or clever: For instance,

```
image_pull_policy_set_to_always {
    image_policy := input.spec.template.spec.containers[_].imagePullPolicy
    image_policy == "Always"
}

violations[rule] {
  input.kind == "Deployment"
  image_pull_policy_set_to_always
  msg := "Image pull policy should usually not be set to 'Always'"

  rule := {
    "ruleID": "0012"
    "message": msg
  }
}
```

Notice how we return an object here with an identifier for a given rule.

**Comment by srenatus** [comment]
I've been thinking about something along those lines, switching the json reader, too... 🤔 Do you know of such a JSON reader? I don't think I've come across one. It could be useful on many occasions...

Re: your snippet, if you try something like this, you'd be able to help pinpoint the offending section:

```rego
# Partial rule: will generate an array of container names that have imagePullPolicy: "Always"
image_pull_policy_set_to_always[container.name]{
    container := input.spec.template.spec.containers[_]
    container.imagePullPolicy == "Always"
}

violations[rule] {
  input.kind == "Deployment"
  image_pull_policy_set_to_always[container_name]
  msg := sprintf("container %s: Image pull policy should usually not be set to 'Always'", [container_name])

  rule := {
    "ruleID": "0012",
    "message": msg
  }
}
```

(I've played with that here, but be aware that the input structure is something I've partly made up on the spot to illustrate the example: https://play.openpolicyagent.org/p/9OAUtRWWWs)

Do you think this is a viable workaround? 😃

**Comment by Lawouach** [comment]
Hey @srenatus, this is a neat trick and I'll be folding it into our OPA for sure.

However, I don't think that will be enough. We use Kustomize to build our entire Kubernetes asset and that generate one heck of a big YAML file. So, while for this example, the container name could be enough, we have other rules that wouldn't have such a token or could happen in so many instances, this would not help much.

I don't know of any json loader in golang that would keep line numbers (I'm not a golang dev by trade, Python here ;)). I doubt it's common.

I guess if the input for OPA is just a generic interface, maybe then I could simply read the file myself in a way that remains valid with what OPA expects? Mind you, I don't think OPA would still let me access that location even if If fed it to it, right?

I think this capability would be rather neat for some scenarios we are building but I can appreciate it's not a priority of course as it's likely quite some work.

**Comment by srenatus** [comment]
Heya!

Yeah, so, my curiosity about a json lib like that is not strictly related to how OPA does it. For OPA, having `input` (or pieces of `data`) come from a file is just one case; most generally, there's nothing to tie a piece of information from input/data to some file location. However, `conftest`'s mode of operation is more file-focused, so I could see a feature like that working very well in conftest.

All that said, you could expand the "trick" I've shared above to include something more helpful to locate your offending setting. You've mentioned jsonpath: While it's not entirely simple to generate a jsonpath representation of some reference, you know more about the location in your rego snippet. You could generate that jsonpath string yourself, like (⚠️ I haven't tried this snippet, just trying to sketch the idea)


```rego
image_pull_policy_set_to_always[[container.name, i]]{
    container := input.spec.template.spec.containers[i]
    container.imagePullPolicy == "Always"
}

violations[rule] {
  input.kind == "Deployment"
  image_pull_policy_set_to_always[[container_name, i]]
  msg := sprintf("container %s: Image pull policy should usually not be set to 'Always'", [container_name])

  rule := {
    "ruleID": "0012",
    "message": msg,
    "jsonpath": sprintf("spec.template.spec.containers[%d].imagePullPolicy", i)
  }
}
```

(The "not entirely simple" approach hinted at above is using `walk()`, but I'm afraid it'll complicate your rego code quite a bit.)

❓ Related to this, would you think a new builtin like `jsonpath(ref: Any) -> string` would be helpful? I'd imagine it would do something along those lines:
```rego
jsonpath(input) == "input"
jsonpath(input.foo.bar.baz[1]) == "input.foo.bar.baz[1]" with input.foo.bar.baz as [1,2,3,4]
```

**Comment by Lawouach** [comment]
Ha Stephan, you're marvellous!

A jsonpath function would clearly help and might be even more robust than line number actually! I can't wait to try your snippet and see if I can get somewhere.

**Comment by srenatus** [comment]
👉 https://github.com/open-policy-agent/opa/issues/2940 I've tried to summarize the `jsonpath` builtin to be added here.

It's nighttime in the US, so this hasn't been discussed yet; but please review the issue and add it it as you see fit. 😃

**Comment by Lawouach** [comment]
Brilliant @srenatus .

I was thinking about this issue. Jsonpath will be awesome and tremendously useful. I still think I'll need line numbers nonetheless down the road.

> However, conftest's mode of operation is more file-focused, so I could see a feature like that working very well in conftest.

Would you suggest we move this discussion overthere then? I'm still interested in running as a programmatic library, rather than a full blown process.

**Comment by srenatus** [comment]
We've discussed this internally yesterday, and it's fine here. As it turned out, adding a `jsonpath` function has similar difficulties to tracking source location of input/data files... So, let's keep this issue around. :+1:

**Comment by Lawouach** [comment]
Thanks for the feedback! I'll keep an eye on this one as it's quite an useful one :)
